### PR TITLE
fix: add request queue to handle concurrent zoom & offset requests

### LINKED ISSUE #18: 🐛 [BUG] - Visual Editor : Concurrent update requests on the zoom and offset
### Description

When moving the canvas in the visual editor, concurrent requests to update the zoom and offset are leading to 403 errors on the API side. 

### Reproduction steps

```bash
1. Go to 'Visual Editor' on the left sidebar
2. Open the inspector and select the network tab
3. Move the canvas (drag'n'n-drop) and zoom
```

### Logs

_No response_

### Browsers

_No response_

### OS

_No response_

### PR DESCRIPTION
**Pull request by AuraOfDivinity**
# Motivation

This PR addresses the 403 Forbidden error that occurs when multiple concurrent requests are sent to update a category. The issue arises because each concurrent request generates a new CSRF token, which invalidates the previous one, causing subsequent requests to fail with a 403 error.

To resolve this, I implemented a request queue that ensures update requests are processed sequentially rather than concurrently. By queueing the requests, we prevent multiple CSRF tokens from being generated simultaneously. This allows each request to complete successfully before the next one is processed, eliminating the token invalidation issue.

Fixes #18 

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (Storybook)
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by marrouchi — CHANGES_REQUESTED** [review]
Woah ! you rock man 🎸 

This issue with this approach is still that it would send multiple subsequest updates.

I was thinking about a solution that would optimize that, for example have a generic hook that we can use elsewhere : 

```ts

type UpdateFunction<T> = (updatedFields: Partial<T>) => Promise<void>;

function useDebouncedUpdate<T>(
  apiUpdate: (data: Partial<T>) => Promise<void>,
  delay: number = 300
): [UpdateFunction<T>] {
  // Ref to store the latest updates
  const updatesRef = useRef<Partial<T>>({});

  // Create a debounced version of the API update function
  const debouncedUpdate = useRef(
    debounce(async () => {
      const updates = { ...updatesRef.current };
      try {
        await apiUpdate(updates);
        // Optionally clear updates after successful API call
        updatesRef.current = {};
      } catch (error) {
        console.error('Debounced API update failed:', error);
        // Optionally handle errors (e.g., retry logic, user notifications)
      }
    }, delay)
  ).current;

  // Updater function to be returned by the hook
  const update = useCallback(
    async (updatedFields: Partial<T>) => {
      // Merge new updates with existing ones
      updatesRef.current = { ...updatesRef.current, ...updatedFields };
      // Trigger the debounced API call
      debouncedUpdate();
    },
    [debouncedUpdate]
  );

  // Cleanup on unmount
  useEffect(() => {
    return () => {
      debouncedUpdate.cancel();
    };
  }, [debouncedUpdate]);

  return [update];
}

export default useDebouncedUpdate;
```

**Comment by marrouchi** [comment]
@AuraOfDivinity please feel free to join our Discord and DM me : https://discord.gg/DXGqQVZq

We really need someone like you ! 👏

**Review by marrouchi — APPROVED** [review]

**Review by marrouchi — COMMENTED** [review]
Thanks @AuraOfDivinity 🙏 Very much appreciated !

File: frontend/src/hooks/useDebouncedUpdate.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { debounce } from "@mui/material";
```
Comment (yassinedorbozgithub) [inline frontend/src/hooks/useDebouncedUpdate.tsx:1]: Hi @AuraOfDivinity, can please add the licence at the top of this file 
```suggestion
/*
 * Copyright © 2024 Hexastack. All rights reserved.
 *
 * Licensed under the GNU Affero General Public License v3.0 (AGPLv3) with the following additional terms:
 * 1. The name "Hexabot" is a trademark of Hexastack. You may not use this name in derivative works without express written permission.
 * 2. All derivative works must include clear attribution to the original creator and software, Hexastack and Hexabot, in a prominent location (e.g., in the software's "About" section, documentation, and README file).
 * 3. SaaS Restriction: This software, or any derivative of it, may not be used to offer a competing product or service (SaaS) without prior written consent from Hexastack. Offering the software as a service or using it in a commercial cloud environment without express permission is strictly prohibited.
 */

import { debounce } from "@mui/material";

```

**Review by yassinedorbozgithub — CHANGES_REQUESTED** [review]

**Pull request merged**